### PR TITLE
CONSOLE-3383: Minimal Dynamic Plugin for the CronTab CRD (MVP)

### PR DESCRIPTION
**Pull request by jhadvig**
This PR contains a MVP for crontab dynamic plugin. It's adding:
- custom list page with custom action provider (action button / kebab)
- yaml template for crontab CR
- i18n support
- README

List page screen:
Screenshot 2023-05-17 at 11 27 56

Create page:
Screenshot 2023-05-17 at 11 27 45

/assign @spadgett 

FYI: I've preserved whole commit history. 
CCing: @raspbeep

### CODE REVIEW

File: Dockerfile
Comment (spadgett) [inline Dockerfile:1]: Let's update the Dockerfile to what we use in the console plugin template

File: charts/crontab-plugin/Chart.yaml
Comment (spadgett) [inline charts/crontab-plugin/Chart.yaml:3]: Let's give this a better description.

File: console-extensions.json
Comment (spadgett) [inline console-extensions.json:14]: We should make sure the nav extension is only enabled when the feature flag is set.

File: i18n-scripts/lexers.js
Comment (spadgett) [inline i18n-scripts/lexers.js:1]: We should consider moving this into the template.

File: locales/en/plugin__crontab-plugin.json
Comment (spadgett) [inline locales/en/plugin__crontab-plugin.json:2]: Should the message keys be empty like this?

File: oc-manifest.yaml
Comment (spadgett) [inline oc-manifest.yaml:1]: Do we need both the manifest and the Helm chart? I would prefer just having one.

File: package.json
Comment (spadgett) [inline package.json:64]:
```suggestion
    "description": "Sample plugin for the OpenShift Console. This plugin adds basic functionality that allows users to manage CronTab custom resources.",
```

File: package.json
Comment (spadgett) [inline package.json:63]:
```suggestion
    "displayName": "Sample CronTab Plugin",
```

File: src/models/CronTabModel.ts
Comment (spadgett) [inline src/models/CronTabModel.ts:23]: It would be better to define a more strict type here.

File: src/models/CronTabModel.ts
Comment (spadgett) [inline src/models/CronTabModel.ts:25]: We intended to remove the concept of resource references from the SDK... Where is this needed?

File: src/models/CronTabModel.ts
Comment (spadgett) [inline src/models/CronTabModel.ts:5]: I think this should be defined using the  model metadata extension. We shouldn't need to reference the model directly in the API IIRC, so we hopefully we can remove this.

https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md#consolemodel-metadata

File: src/utils/utils.ts
Comment (spadgett) [inline src/utils/utils.ts:8]: It would be good to try to remove these. We don't want plugins to need to do these kinds of transformations.

File: src/views/CronTabList/CronTabList.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:46]: Should `groupVersionKind` be the object, not the reference string? If that's not the case, we need to open a bug against the SDK.

File: src/views/CronTabList/CronTabList.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:66]: Do we need these width classes? By default the columns should all be the same width?

File: src/views/hooks/useCronTabActionsProvider.tsx
Comment (spadgett) [inline src/views/hooks/useCronTabActionsProvider.tsx:23]: Avoid manually building URLs if at all possible.

**Review by spadgett — COMMENTED** [review]
Thanks @jhadvig

File: oc-manifest.yaml
Comment (jhadvig) [reply oc-manifest.yaml:1]: The motivation here was that not everybody is using Helm, and would rather prefer using `oc`

File: src/models/CronTabModel.ts
Comment (jhadvig) [reply src/models/CronTabModel.ts:25]: this was used when manually building the URLs

File: src/views/components/CronTabActions/CronTabActions.tsx
Comment (jhadvig) [reply src/views/components/CronTabActions/CronTabActions.tsx:33]:
```suggestion
  const handleClick = (action: Action) => {
    if (typeof action?.cta === 'function') {
      action.cta();
    } else {
      console.error('CronTabActions: action.cta is not a function');
    }
    setIsOpen(false);
  };
```

Comment (jhadvig) [reply src/views/components/CronTabActions/CronTabActions.tsx:33]: added following condition

File: src/views/components/CronTabActions/CronTabActions.scss
Comment (jhadvig) [reply src/views/components/CronTabActions/CronTabActions.scss:5]: not needed... removed it

File: src/views/components/CronTabActions/CronTabActions.scss
Comment (jhadvig) [reply src/views/components/CronTabActions/CronTabActions.scss:7]: removed

File: src/views/CronTabList/CronTabList.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabList.tsx:91]: removed

File: src/views/CronTabList/CronTabList.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabList.tsx:66]: removed

File: src/views/CronTabList/CronTabList.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabList.tsx:46]: Unfortunately `groupVersionKind` is type of [GroupVersionKind](https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/api/dynamic-core-api.ts#L177) which is string.
We need to update to K8sGroupVersionKind which is an object

File: i18n-scripts/lexers.js
Comment (jhadvig) [reply i18n-scripts/lexers.js:1]: not sure what template you mean?

File: locales/en/plugin__crontab-plugin.json
Comment (jhadvig) [reply locales/en/plugin__crontab-plugin.json:2]: no it does not.. hmm not sure whats happening since I have `useKeysAsDefaultValue: true,` in my `i18next-parser.config.js`

Comment (jhadvig) [reply locales/en/plugin__crontab-plugin.json:2]: also `yarn i18n` scripts passes without any issue and creates the locale file :-/

File: i18n-scripts/lexers.js
Comment (spadgett) [reply i18n-scripts/lexers.js:1]: openshift/console-plugin-template

Comment (jhadvig) [reply i18n-scripts/lexers.js:1]: will create a story for that

File: src/views/CronTabList/CronTabList.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabList.tsx:46]: Created https://issues.redhat.com/browse/OCPBUGS-13808

File: src/models/CronTabModel.ts
Comment (jhadvig) [reply src/models/CronTabModel.ts:25]: So Im not sure how to avoid building the url manually at this point.
I start to think this is a tech debt on our side and we should expose a function which will help plugins creators with crafting URLs.
Thoughts?

Comment (spadgett) [reply src/models/CronTabModel.ts:25]: Yeah, I think we need an SDK method for building resource URLs. Then if we ever need to change how the URLs are built (e.g. if we remove the API resource version from the URL) plugins aren't broken.

**Comment by jhadvig** [comment]
@spadgett addressed most of the comments.
Regarding building URL manually and having models part of the codebase, I think we need to address those as a techdebt into the SDK.
Will create a follow up stories for those.

File: oc-manifest.yaml
Comment (jhadvig) [reply oc-manifest.yaml]: removed the `oc-manifest.yaml` and updated the README

File: .gitignore
Comment (spadgett) [inline .gitignore:6]: We don't have any integration tests yet, correct?

File: README.md
Comment (spadgett) [inline README.md:8]:
```suggestion
The CronTab Dynamic Plugin creates a new navigation entry, list page, details page, and YAML template for the CRD.
```

File: charts/crontab-plugin/Chart.yaml
Comment (spadgett) [inline charts/crontab-plugin/Chart.yaml:3]:
```suggestion
description: A Helm chart for the sample CronTab Dynamic Plugin used in OpenShift Console.
```

File: i18n-scripts/common.js
Comment (spadgett) [inline i18n-scripts/common.js:1]: Are we sure we need all of these i18n scripts? This is a lot of code to be copying around.

If they're really needed, we should make this available as a package that plugins can use. But I'm surprised we need this much custom code to use i18next-parser for a simple project.

File: src/models/CronTabModel.ts
Comment (spadgett) [reply src/models/CronTabModel.ts:5]: This comment still stands. I don't think we should be creating a K8sModel resource, even if we have to build a reference somewhere.

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabRowActions.tsx:24]: I'm inclined to remove the actions from the initial PR until we can fix this properly. The problem is people will copy this sample code into their own plugins, and even if we update the sample, those real plugins won't get updated.

I'd suggest we get something working for the current released OpenShift versions and make sure they're in an appropriate release-* branch, then we can come back and add actions when the SDK functions we need are available.

File: src/views/components/CronTabActions/CronTabActions.tsx
Comment (spadgett) [inline src/views/components/CronTabActions/CronTabActions.tsx:55]: Remove data-test-id if we don't have any test code

File: src/views/CronTabList/CronTabList.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:68]: Just use

```
    groupVersionKind: {
      group: 'stable.example.com',
      kind: 'CronTab',
      version: 'v1',
    }
```

and remove the modelTo helper

**Review by spadgett — COMMENTED** [review]
Thanks, @jhadvig

File: i18n-scripts/common.js
Comment (jhadvig) [reply i18n-scripts/common.js:1]: So I was trying to decouple it but the default values havent been generated for the strings, for some reason.

File: src/views/components/CronTabActions/CronTabActions.tsx
Comment (jhadvig) [reply src/views/components/CronTabActions/CronTabActions.tsx:55]: removed

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabRowActions.tsx:24]: Makes sense. On the other side folks will check how other plugins are doing (constructing the URL manually) and use that.

But I like it more then having it here... will remove 👍

File: src/views/CronTabList/CronTabList.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabList.tsx:68]: done

**Comment by jhadvig** [comment]
@spadgett ping

File: src/utils/utils.ts
Comment (spadgett) [inline src/utils/utils.ts:3]: Is this still used?

File: src/views/CronTabList/CronTabList.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:45]:
```suggestion
        {/* groupVeersionKind should be take an object but there is discrepancy in the prop types
        https://issues.redhat.com/browse/OCPBUGS-13808 will update the types */}
```

File: src/views/CronTabList/CronTabList.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:46]:
```suggestion
        <ListPageCreate groupVersionKind="stable.example.com~v1~CronTab">Create CronTab</ListPageCreate>
```

File: src/views/CronTabList/CronTabList.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabList.tsx:68]: nit: I'd pull this out into a const somewhere so we don't repeat it in several places

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabRowActions.tsx:23]: If there are no items, I would leave the dropdown out of the PR for now.

File: src/views/hooks/useCronTabActionsProvider.tsx
Comment (spadgett) [inline src/views/hooks/useCronTabActionsProvider.tsx:14]: If there are no actions, I'd leave this out. We can keep this in a separate branch so we still have the code.

File: src/utils/utils.ts
Comment (jhadvig) [reply src/utils/utils.ts:3]: removed

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabRowActions.tsx:23]: wanted to preserve, if it would be used as a starting point for a new plugin, but can remove

File: src/views/hooks/useCronTabActionsProvider.tsx
Comment (jhadvig) [reply src/views/hooks/useCronTabActionsProvider.tsx:14]: removed

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (spadgett) [inline src/views/CronTabList/CronTabRowActions.tsx:1]: I would take this entire component out from the PR if we have no actions.

We can add the edit crontab modal in a follow up and add this back.

**Comment by jhadvig** [comment]
@spadgett comments addressed :)

File: src/views/CronTabList/CronTabRowActions.tsx
Comment (jhadvig) [reply src/views/CronTabList/CronTabRowActions.tsx:1]: removed

**Review by spadgett — COMMENTED** [review]
LGTM, can we clean up the commit history a bit before merging? It doesn't have to be a single commit, but would be nice to squash things to remove "WIP" and "addressing comments" type commits.

**Comment by jhadvig** [comment]
For sure ! was just waiting for LGTM 😅

File: src/views/components/CronTabActions/CronTabActions.tsx
Comment (spadgett) [inline src/views/components/CronTabActions/CronTabActions.tsx:1]: Not sure if I missed on previous review, but we should remove until we have actions.

**Comment by jhadvig** [comment]
@spadgett comment addressed. PTAL

**Review by spadgett — APPROVED** [review]
/lgtm

**Pull request merged**
